psi4/src/psi4/libfock/DirectJK.cc

```diff
@@ -412,19 +420,20 @@ void DirectJK::compute_JK() {
             ints.push_back(std::shared_ptr<TwoBodyAOInt>(ints[0]->clone()));
         }
         if (do_J_ && do_K_) {
-            build_JK(ints, D_ref, J_ref, K_ref);
-        } else if (do_J_) {
-            std::vector<std::shared_ptr<Matrix>> temp;
-            for (size_t i = 0; i < D_ao_.size(); i++) {
-                temp.push_back(std::make_shared<Matrix>("temp", primary_->nbf(), primary_->nbf()));
+            if (!linK_) {
+                build_JK_matrices(ints, D_ref, J_ref, K_ref);
+            } else {
+                build_linK(ints, D_ref, K_ref);
```

Comment:
Yes, this is correct. LinK will not be useful until CompositeJK, with the Direct-DF-J algorithm has been merged into Psi4

Reply:
Did you misspeak, or did you say this PR is _not going to be useful_ before other algorithms are implemented?

Reply:
With this PR, users can run LinK, and it will work perfectly fine, BUT, there will be no expected performance gain from it, due to integral reuse in the Direct algorithm. This PR is really more meant to be a "stepping stone PR", where we verify the correctness of LinK first. However, LinK is INDEED useful when combined with Direct-DF-J or CFMM, and I have personally verified this with timings. Combining Direct-DF-J, LinK, as well as a new JK class where we split up J and K would be a behemoth of a PR. Maybe I should add in the docs that LinK is currently recommended only for research purposes and not for performance?

Reply:
> With this PR, users can run LinK, and it will work perfectly fine, BUT, there will be no expected performance gain from it, due to integral reuse in the Direct algorithm.

In the introduction to this PR, you claimed this PR adds
* Faster build of K matrix based on pre-screening techniques
* Verified timings and accuracy of calculations on 3D system (watercluster)

And I'm struggling to reconcile those statements with "no expected performance gain." Can you elaborate on the timing benchmarks you have on `LinK`?

> This PR is really more meant to be a "stepping stone PR", where we verify the correctness of LinK first. However, LinK is INDEED useful when combined with Direct-DF-J or CFMM, and I have personally verified this with timings. Combining Direct-DF-J, LinK, as well as a new JK class where we split up J and K would be a behemoth of a PR.

Splitting it like this was the right decision.

> Maybe I should add in the docs that LinK is currently recommended only for research purposes and not for performance?

I insist on it. Also, please double-check that `LinK` is not used by default in this PR, in cases where we can predict an efficiency loss.

Reply:
> In the introduction to this PR, you claimed this PR adds
> 
> * Faster build of K matrix based on pre-screening techniques
> * Verified timings and accuracy of calculations on 3D system (watercluster)
> 
> And I'm struggling to reconcile those statements with "no expected performance gain." Can you elaborate on the timing benchmarks you have on `LinK`?

Yes, the timings I ran were comparing only the K build portion of the DirectJK algorithm when LinK is turned on to the run time for running DirectJK (building J and K together in the integral-direct loop). LinK is lightning-fast compared to that. The reason why there is no performance gain currently from LinK is because J has to be built as well, and the current algorithm used to build J with LinK is the plain-old integral-direct J algorithm. This was what inspired my work on Direct-DF-J and CFMM, since the only way for a linear scaling K algorithm to be useful is if it is combined with a fast J algorithm.